src/server/data_source.rs

```diff
@@ -1,142 +1,197 @@
-
-/// Write callback for [`DataSource`].
 #[allow(clippy::module_name_repetitions)]
-pub type DataSourceWrite = dyn FnMut(&mut DataSourceWriteContext) -> DataSourceResult;
+pub type DataSourceResult = std::result::Result<(), ua::StatusCode>;
 
 /// Data source with callbacks.
 ///
-/// The read and write callbacks implement the operations on the data source when it is added as a
-/// variable node via [`Server::add_data_source_variable_node()`].
+/// The read and write callbacks implement the operations on the variable when it is added via
+/// [`Server::add_data_source_variable_node()`].
 ///
 /// [`Server::add_data_source_variable_node()`]: crate::Server::add_data_source_variable_node
-pub struct DataSource {
-    read: Box<DataSourceRead>,
-    write: Option<Box<DataSourceWrite>>,
+pub trait DataSource {
+    /// Reads from variable.
+    ///
+    /// This is called when a client wants to read the value from the variable.
+    ///
+    /// # Errors
+    ///
+    /// This should return an appropriate error when the read is not possible. The underlying status
+    /// code is forwarded to the client.
+    fn read(&mut self, context: &mut DataSourceReadContext) -> DataSourceResult;
+
+    /// Writes to variable.
+    ///
+    /// This is called when a client wants to write the value to the variable.
+    ///
+    /// # Errors
+    ///
+    /// This should return an appropriate error when the write is not possible. The underlying
+    /// status code is forwarded to the client.
+    fn write(&mut self, context: &mut DataSourceWriteContext) -> DataSourceResult;
 }
 
-impl DataSource {
-    /// Defines read-only data source.
-    #[must_use]
-    pub fn read_only(
-        read: impl FnMut(&mut DataSourceReadContext) -> DataSourceResult + 'static,
-    ) -> Self {
-        Self {
-            read: Box::new(read),
-            write: None,
-        }
+/// Context when [`DataSource`] is being read from.
+#[allow(clippy::module_name_repetitions)]
+pub struct DataSourceReadContext {
+    /// Outgoing value to be read.
+    ///
+    /// This is a mutable cell where the read callback puts the data to be returned to the client.
+    value_target: NonNull<UA_DataValue>,
+}
+
+impl DataSourceReadContext {
+    /// Creates context for read callback.
+    fn new(value: *mut UA_DataValue) -> Option<Self> {
+        Some(Self {
+            value_target: NonNull::new(value)?,
+        })
     }
 
-    /// Defines writable data source.
-    #[must_use]
-    pub fn read_write(
-        read: impl FnMut(&mut DataSourceReadContext) -> DataSourceResult + 'static,
-        write: impl FnMut(&mut DataSourceWriteContext) -> DataSourceResult + 'static,
-    ) -> Self {
-        Self {
-            read: Box::new(read),
-            write: Some(Box::new(write)),
-        }
+    /// Sets value.
+    ///
+    /// This sets the value to report back to the client that is reading from this [`DataSource`].
+    pub fn set_value(&mut self, value: ua::DataValue) {
```

Comment:
What happens if `set_value()` is called twice with different values? Do both contexts obtain ownership on the wrapped values?